src/Embedders/Word2Vec.php

```diff
@@ -0,0 +1,983 @@
+		int $dimensions   = 5,
+		float $sampleRate = 1e-3,
+		float $alpha      = 0.01,
+		int $epochs       = 10,
+		int $minCount     = 2
+	) {	
+
+		if(!$this->acceptedLayer($layer)){
+			throw new InvalidArgumentException('Layer must be neg or hs.');
+		}
+
+		if($window > 5){
+			throw new InvalidArgumentException('Window must be less than 5, $window given.');
+		}
+
+        if ($dimensions < 5) {
+            throw new InvalidArgumentException("Dimensions must be greater than 4, $dimensions given.");
+        }
+
+        if ($sampleRate < 0.0) {
+            throw new InvalidArgumentException("Sample rate must be 0 or greater, $sampleRate given.");
+        }
+
+        if ($alpha < 0.0) {
+            throw new InvalidArgumentException("Alpha must be greater than 0, $alpha given.");
+        }
+
+        if ($epochs < 1) {
+            throw new InvalidArgumentException("Number of epochs must be greater than 0, $epochs given.");
+        }	
+
+        if ($minCount < 1) {
+            throw new InvalidArgumentException("Minimum word count must be greater than 0, $minCount given.");
+        }	        	
+
+		$this->layer 	    = $layer;
+		$this->window 	    = $window;		
+		$this->dimensions 	= $dimensions;
+		$this->sampleRate   = $sampleRate;
+		$this->alpha        = $alpha;		
+		$this->epochs     	= $epochs;
+		$this->minCount  	= $minCount;
+		$this->neg_labels 	= Vector::quick([1, 0]);
+	}
+
+    /**
+     * Return the data types that this estimator is compatible with.
+     *
+     * @return \Rubix\ML\DataType[]
+     */
+    public function compatibility() : array
+    {
+        return [
+            DataType::categorical(),
+        ];
+    }	
+	
+    /**
+     * Return the settings of the hyper-parameters in an associative array.
+     *
+     * @return mixed[]
+     */
+    public function params() : array
+    {
+        return [
+            'layer'       => $this->layer,
+            'window'      => $this->window,
+            'dimensions'  => $this->dimensions,
+            'sample_rate' => $this->sampleRate,
+            'alpha'       => $this->alpha,
+            'epochs'      => $this->epochs,
+            'min_count'   => $this->minCount,
+        ];
+    }
+
+    /**
+     * Has the learner been trained?
+     *
+     * @return bool
+     */
+    public function trained() : bool
+    {	
+    	if(empty($this->vectors)) {
+    		return false;
+    	}
+
+    	return true;
+    }
+
+    /**
+     * Iterating through a range of the specific epoch count and updating all respective word vectors.  
+     * @param \Rubix\ML\Datasets\Dataset $dataset
+     * @throws \InvalidArgumentException
+     *
+     */		
+	public function train(Dataset $dataset) : void	
+	{
+		DatasetIsNotEmpty::check($dataset);
+
+		$sentences = $dataset->column(0);
+
+		$this->preprocess($sentences);
+		$this->prepareWeights();
+
+		$min_alpha   = .0001;
+		$start_alpha = $this->alpha;
+
+		foreach(range(0, ( $this->epochs - 1)) as $i){	
```

Comment:
```foreach (range(...) as $i) {``` works differently in Python than it does in PHP ... This line will actually create and allocate memory for the array produced by range which makes it slower and less efficient than ```for ($i = 0; $i < $this->epochs; ++$i) {```.

Refer this comment https://www.php.net/manual/en/function.range.php#85331